### ISSUE TITLE
Runtime type signature/annotations apparently erased when inheriting from generics?

### ISSUE THREAD

**Issue by Fidget-Spinner** [python/cpython#94897]
During EuroPython @elfjes told me that runtime annotations are erased when inheriting from generics.

Sorry but right now I'm not too sure about what you meant. So could you please create a reproducer for me? Then I can proceed with something.

**Comment by elfjes** [comment]
Hi @Fidget-Spinner, thanks for talking with me at EP and having a look at this issue. 

I had been a while since I encountered issues with `typing.Generic` and as I'm writing this, I see that these issues have been (mostly) resolved.

The code I'm working deals with dependency injection and automatic resolution of class constructor dependencies. It resolves around the objective: 

"given a type-annotated class/`__init__`, give me the argument names and (concrete) types of that class' constructor argument so that I can create an instance of that class". 

For simple annotated classes, this is easy, just look at either `Cls.__init__.__annotations__` or `inspect.signature(Cls)`. `inspect.signature` is the better choice here, because it also provides information about variadic positional and keyword argument. There is one major drawback of `inspect.signature`, in that it cannot properly determine the signature when `__new__` has been overridden  (with ie `__new__(cls, *args, **kwargs)`). In that case `inspect.signature` returns the signature of `__new__` (although I realize now that `inspect.signature(Cls.__init__)` still provides the necessary information, so I guess that solves that issue)

Now, previously `typing.Generic` did override `__new__`, leading to the above issue, but since 3.9 it doesn't anymore. So that's no longer an issue (regardless of the above troubles with `inspect.signature`).

I still have to properly solve for all the different cases when dealing with `__origin__` to determine the concrete type underlying generics, for which `typing.Generic` is also an interesting case. Consider the following:

```python
T = typing.TypeVar("T")

class GenericDependency(typing.Generic[T]):
    pass

class DependingConcrete:
    def __init__(self, dep: GenericDependency):
        # the type hint for `dep` is the concrete class GenericDependency
        pass


class DependingGeneric(typing.Generic[T]):
    def __init__(self, dep: GenericDependency[T]):
        # to find the concrete class of `dep` we need to look at the type hint's `__origin__`, which is kind of cumbersome, 
        # and may easily conflict with other logic that deals with `__origin__` on other classes like the `_GenericAlias` ones,
        # but not insurmountable
        pass
```

Although I also see now that with newer versions of python `typing.get_origin` has also been much improved, so I can also check out to see if that function can help me with the whole `__origin__` story.

So long story short, I think the troubles I was facing are mostly resolved and this issue can be closed. Thanks for your willingness to help anyway ^_^

**Comment by Fidget-Spinner** [comment]
Yay, that's great news! No problem.

BTW, I really recommend using [typing.get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) or [inspect.get_annotations](https://docs.python.org/3/library/inspect.html#inspect.get_annotations) (only in >=3.10) if you want to grab annotations instead of using `inspect.signature`. The main difference between the two is that `get_annotations` doesn't evaluate string annotations (forwardrefs) while `get_type_hints` does.

**Comment by elfjes** [comment]
Yes, I forgot to mention that I'm also using `typing.get_type_hints`, especially for resolving forwardrefs. It needs to be used however together with `inspect.signature` since there might be arguments without a type hint, and to figure out if any of the arguments are variadic.
